File: acwing/蓝桥杯每日一题2025/week1/6134.哞叫时间II/6134.哞叫时间II.cpp
#include <iostream>
#include <algorithm>
#include <cstring>
#include <cstdio>

using namespace std;

const int N = 1e6 + 10;
typedef long long LL;

int a[N], l[N], r[N], cnt;

int main()
{
    int n;
    scanf("%d", &n);
    for (int i = 1; i <= n; i++)
        scanf("%d", &a[i]);
    for (int i = 1; i <= n; i++)
    {
        int x = a[i];
        if (++l[x] == 1)
            cnt++;
    }
    LL res = 0;
    for (int i = n; i >= 1; i--)
    {
        int x = a[i];
        r[x] += 1;
        l[x] -= 1;
        if (l[x] == 0)
            cnt -= 1;
        if (r[x] == 2)
            res += cnt - (l[x] > 0 ? 1 : 0);
    }
    printf("%lld", res);
    return 0;
}

File: acwing/蓝桥杯每日一题2025/week1/6134.哞叫时间II/6134.哞叫时间II.py
N = int(1e6) + 10
cnt = 0
l = [0] * N
r = [0] * N

n = int(input())
a = list(map(int, input().split()))

for i in range(n):
    x = a[i]
    if l[x] == 0:
        cnt += 1
    l[x] += 1

res = 0
for i in range(n - 1, -1, -1):
    x = a[i]
    r[x] += 1
    l[x] -= 1
    if l[x] == 0:
        cnt -= 1
    if r[x] == 2:
        res += cnt
        if l[x] > 0:
            res -= 1
print(res)
